includes/functions.php

```diff
@@ -624,34 +595,97 @@ function tsml_geocode_google($address, $tsml_map_key)
+	if ($tsml_geocoding_method === 'api_gateway') {
+		curl_setopt($tsml_curl_handle, CURLOPT_URL, 'https://api-gateway.apps.itstechnical.net/api/geocode?' . http_build_query($geocode_parameters));
+	} elseif ($tsml_geocoding_method === 'c4r_geocoding') {
+
+		if (!$registration = get_option('tsml_geocoding_registration')) {
+
+			//create registration request
+			$registration_parameters = [
+				"organizationName" => get_bloginfo('name'),
+				"websiteURL" => get_bloginfo('url'),
+				"metadata" => [
+					[
+						"keyName" => "appName",
+						"value" => "TSML",
+					], [
+						"keyName" => "appVersion",
+						"value" => TSML_VERSION,
+					], [
+						"keyName" => "wordpressVersion",
+						"value" => get_bloginfo('version'),
+					],
+				]
+			];
+
+			//gateway only supports certain program values
+			$supported_programs = ['AA', 'Al-Anon', 'NA', 'CA', 'SA'];
+			if (!empty($tsml_programs[$tsml_program]['abbr']) && in_array($tsml_programs[$tsml_program]['abbr'], $supported_programs)) {
+				$registration_parameters['metadata'][] = [
+					"keyName" => "recoveryProgram",
+					"value" => $tsml_programs[$tsml_program]['abbr'],
+				];
+			}
+
+			curl_setopt($tsml_curl_handle, CURLOPT_URL, "https://geocode.code4recovery.org/api/v1/users");
+			curl_setopt($tsml_curl_handle, CURLOPT_POST, 1);
+			curl_setopt($tsml_curl_handle, CURLOPT_POSTFIELDS, http_build_query($registration_parameters));
+			$result = curl_exec($tsml_curl_handle);
+
+			//connection error
+			if ($result === false) {
+				return 'Could not connect to ' . $provider . ' to register. Response was <code>' . curl_error($tsml_curl_handle) . '</code>';
+			}
+
+			$data = json_decode($result, true);
+
+			//registration error
+			if (empty($data['apiToken'])) {
+				return 'An error occurred when registering with ' . $provider . '. The server returned <code>' . $result . '</code>';
+			}
+
+			//registration was succesful
+			$registration = $registration_parameters + $data + [
+				'registered_at' => current_time('mysql'),
+			];
+
+			//save registration
+			update_option('tsml_geocoding_registration', $registration);
+
+			//turn POST off again 
+			curl_setopt($tsml_curl_handle, CURLOPT_POST, 0);
+		}
+
+		$geocode_parameters['apiToken'] = $registration['apiToken'];
 
-	//send request to google
-	if ($tsml_geocoding_method == 'api_gateway') {
-		curl_setopt($tsml_curl_handle, CURLOPT_URL, 'https://api-gateway.apps.itstechnical.net/api/geocode?' . http_build_query($options));
+		curl_setopt($tsml_curl_handle, CURLOPT_URL, 'https://geocode.code4recovery.org/api/v1/geocodes?' . http_build_query($geocode_parameters));
 	} else {
-		curl_setopt($tsml_curl_handle, CURLOPT_URL, 'https://maps.googleapis.com/maps/api/geocode/json?' . http_build_query($options));
+		//use the user's google key if indicated
+		$geocode_parameters['key'] = ($tsml_geocoding_method === 'google_key' && !empty($tsml_google_maps_key))
```

Comment:
If the user has a Google mapping key configured, and legacy set, this will default to their mapping key for geocoding, right?

Assuming I'm reading that correctly, I don't believe we want that for legacy as their mapping key might not have geocoding enabled as a service.